### PR TITLE
Add Nether generator

### PR DESCRIPTION
**Pull request by jimmikaelkael**
This replaces the old nether generation.
Related issue: #16

### CODE REVIEW

**Comment by jimmikaelkael** [comment]
It's complete. Just missing physics to flow the lava and some nether fortress structures, but for now it's good.

**Comment by mastercoms** [comment]
I'll merge https://github.com/GlowstoneMC/Glowstone/pull/636 this first.

**Comment by mastercoms** [comment]
Tested this. Beautiful. Nice work.

Except, there is random gravel at around Y: 64.
and biomes are not set properly, though it isn't that important.

File: src/main/java/net/glowstone/generator/NetherGenerator.java [hunk 1]
```diff
@@ -0,0 +1,288 @@
...
+    @Override
+    public short[][] generateExtBlockSectionsWithData(World world, Random random, int chunkX, int chunkZ, BiomeGrid biomes) {
+        final short[][] buf = generateRawTerrain(world, chunkX, chunkZ);
+
+        int cx = chunkX << 4;
+        int cz = chunkZ << 4;
+
+        final double[] surfaceNoise = ((PerlinOctaveGenerator) getWorldOctaves(world).get("surface")).fBm(cx, cz, 0, 0.5D, 2.0D);
+        final double[] soulsandNoise = ((PerlinOctaveGenerator) getWorldOctaves(world).get("soulsand")).fBm(cx, cz, 0, 0.5D, 2.0D);
+        final double[] gravelNoise = ((PerlinOctaveGenerator) getWorldOctaves(world).get("gravel")).fBm(cx, 0, cz, 0.5D, 2.0D);
```
Comment (mastercoms) [inline src/main/java/net/glowstone/generator/NetherGenerator.java]: This seems to be generating the gravel, though you probably knew that already :)

**Comment by jimmikaelkael** [comment]
@mastercoms things must have changed, here's what I found in a freshly generated world with a seed:
http://i.imgur.com/kPSRuDr.png

Screenshot taken from vanilla minecraft 1.8.3 in solo.

**Comment by mastercoms** [comment]
Yeah, [you're right](http://minecraft.gamepedia.com/Nether#Naturally_generated). Sorry for not investigating it further.

But this feels wrong to me:
gravel slice in nether

**Comment by jimmikaelkael** [comment]
Like you said before, giving the user more control  on the noises is a how to handle such thing.

**Comment by mastercoms** [comment]
Yes but default values should be good.

Does gravel only spawn near Y: 64, or am I missing something again?

**Comment by jimmikaelkael** [comment]
Yeah, from Y 63 to 65.

**Comment by mastercoms** [comment]
What would happen if that range were to be expanded?

**Comment by mastercoms** [comment]
@deathcap any concerns or suggestions before we merge?

**Comment by jimmikaelkael** [comment]
I don't really undertand your last question... More gravel I guess :)

**Comment by mastercoms** [comment]
Well would those gravel patches appear on a larger variation of heights, with the same thickness or would the gravel patches just get thicker?

**Comment by jimmikaelkael** [comment]
Same tickness on larger height. It's just a noise taken at XZ, if noise treshold is reach Y is between 63 and 65 it places gravel and there's netherrack, but Y isn't influencing this noise.

**Comment by mastercoms** [comment]
Ok. So is there a way to expand the range but decrease the probability of gravel?

File: src/main/java/net/glowstone/generator/NetherGenerator.java (hunk 1 above)
Comment (jimmikaelkael) [inline src/main/java/net/glowstone/generator/NetherGenerator.java]: There, 0.5 is the frequency and 2.0 the amplitude of the fractal.

**Comment by mastercoms** [comment]
I'll test some new values and get back to you, then.

**Comment by jimmikaelkael** [comment]
For expanding the range it's required to modify the loop that build terrain columns.

**Comment by mastercoms** [comment]
So amplitude cannot be increased on its own?

**Comment by jimmikaelkael** [comment]
Amplitude will still influence the fractional brownian motion, but it's 2d noise so Y isn't used. 
But by range I concluded you meant height ?
Maybe I was wrong...

**Comment by mastercoms** [comment]
Ah, I misunderstood what you meant by amplitude.

**Comment by mastercoms** [comment]
I expanded the height range, but I couldn't get the frequency right, it was either too high or too low, so I'm giving up until tomorrow.

File: src/main/java/net/glowstone/generator/NetherGenerator.java [hunk 2]
```diff
@@ -0,0 +1,288 @@
...
+        octaves.put("surface", gen);
+
+        gen = new PerlinOctaveGenerator(seed, 4, 16, 16, 1);
+        gen.setXScale(SURFACE_SCALE / 2.0);
+        gen.setYScale(SURFACE_SCALE / 2.0);
+        octaves.put("soulsand", gen);
+
+        gen = new PerlinOctaveGenerator(seed, 4, 16, 1, 16);
+        gen.setXScale(SURFACE_SCALE / 2.0);
+        gen.setZScale(SURFACE_SCALE / 2.0);
```
Comment (jimmikaelkael) [inline src/main/java/net/glowstone/generator/NetherGenerator.java]: The noise scales can be played with too.

**Comment by jimmikaelkael** [comment]
As commented above you can play with noise scales too. Keep in mind that this noise is not very detailed has it only have 4 octaves. More octaves would give more details but less performances.

**Pull request merged**